lewis/adapters/stream.py

```diff
@@ -163,8 +186,12 @@ def __init__(self, func, pattern, argument_mappings=None, return_mapping=None, d
             raise RuntimeError('Can not construct a Func-object from a non callable object.')
 
         self.func = func
-        self.raw_pattern = pattern
-        self.pattern = re.compile(b(pattern), 0) if pattern else None
+
+        self.raw_pattern = pattern.raw_pattern
+        self.pattern = re.compile(b(pattern.regex_pattern), 0)
```

Comment:
I like the helper classes for explicitly differentiating between regex and scanf, but I'm still in favour of some kind of auto-detection over a per-interface default via attribute... We discussed it before, but now that there's some code to look at and consider, I'll try to sell you the idea one more time: :car:

For the device developer, it probably won't make a huge difference either way (although I do think auto-detection is a bit more intuitive / convenient) since they have to learn use whatever tools we give them, and both options get the job done. But I think it makes a huge difference on implementation complexity (and as a result, maintainability) for us here.

Consider if, right above here, we insert these two lines:
```python
if isinstance(pattern, string_types):
    pattern = regex(pattern) if pattern.startswith('^') else fmt(pattern)
```
With this, Lewis would auto-detect between `regex` and `scanf` based on a very simple (to learn and remember) rule when `pattern` is a string type. You could switch freely within the same device without needing to explicitly specify `fmt('...')` or `regex('...')`. No need to define or know that you need to override a `pattern_type` attribute to get `fmt` by default. You would still be able to specify the format type explicitly if you wanted to, but it would almost never be needed (really only if you need a `^` at the beginning of a `fmt`).

Most importantly: ALL of the other changes in this file can go away! We just need the two format helper classes and this block in `Func.__init__`. Instead feature being split across 5 different functions in 5 different classes, it would implemented in one constructor of this one class (not counting `fmt` and `regex` in both cases). This makes understanding and maintaining how the feature works much easier, since it is much more localized.

Command definitions would look like this:
```python
    commands = {
        Cmd('implicit_regex', r'^T\?$'),
        Cmd('implicit_scanf', r'T=%f''),
        Cmd('explicit_regex', regex(r'^P\?$')),  # Should never be needed, since you should always start with ^
        Cmd('explicit_scanf', fmt('^P=%f')),  # Only ever needed if a really weird device starts commands with ^
    }
```

Would there be a disadvantage to doing it this way?